src/System.Private.CoreLib/shared/System/Marvin.cs

```diff
@@ -84,7 +83,86 @@ public static int ComputeHash32(ref byte data, int count, ulong seed)
                     goto case 3;
 
                 case 3:
-                    p0 += 0x80000000u | (((uint)(Unsafe.AddByteOffset(ref data, byteOffset + 2))) << 16)| (uint)(Unsafe.ReadUnaligned<ushort>(ref Unsafe.AddByteOffset(ref data, byteOffset)));
+                    p0 += 0x80000000u | (((uint)(Unsafe.AddByteOffset(ref data, byteOffset + 2))) << 16) | (uint)(Unsafe.ReadUnaligned<ushort>(ref Unsafe.AddByteOffset(ref data, byteOffset)));
+                    break;
+
+                default:
+                    Debug.Fail("Should not get here.");
+                    break;
+            }
+
+            Block(ref p0, ref p1);
+            Block(ref p0, ref p1);
+
+            return (int)(p1 ^ p0);
+        }
+
+        /// <summary>
+        /// Compute a Marvin OrdinalIgnoreCase hash and collapse it into a 32-bit hash.
+        /// n.b. <paramref name="count"/> is specified as char count, not byte count.
+        /// </summary>
+        public static int ComputeHash32OrdinalIgnoreCase(ref char data, int count, uint p0, uint p1)
+        {
+            nuint ucount = (nuint)count;
+            nuint byteOffset = 0;
+            uint tempValue;
+
+            while (ucount >= 4)
+            {
+                tempValue = Unsafe.ReadUnaligned<uint>(ref Unsafe.As<char, byte>(ref Unsafe.AddByteOffset(ref data, byteOffset)));
+                if (!Utf16Utility.AllCharsInUInt32AreAscii(tempValue))
+                {
+                    goto NotAscii;
+                }
+                p0 += Utf16Utility.ConvertAllAsciiCharsInUInt32ToUppercase(tempValue);
+                Block(ref p0, ref p1);
+
+                tempValue = Unsafe.ReadUnaligned<uint>(ref Unsafe.As<char, byte>(ref Unsafe.AddByteOffset(ref data, byteOffset + 4)));
```

Comment:
Is there measurable benefit from manually unrolling this look twice?

Reply:
This is a direct port of the regular Marvin logic. I didn't think it would be appropriate to change the overall structure as part of this PR. We could consider it, though.

On a related note, I _did_ test changing the regular Marvin logic from two unaligned 32-bit reads to a single unaligned 64-bit read (on 64-bit platforms). That resulted in a 5%-ish perf improvement on the inner loop. I've also been experimenting with various other optimizations but didn't want to bundle them with this specific PR.

Reply:
The original Marvin impl had the manual unrolling done based on measurements that proved that it improves performance significantly (>10%).

This change is adding a bunch more code to the loop and so it should be re-measured whether the manual unrolling is still worth it. I do not think that removing the manual unrolling would qualify as significant structure change (it is just deleting code and changing a constant).